va/dns.go

```diff
@@ -91,3 +89,86 @@ func (va *ValidationAuthorityImpl) validateDNS01(ctx context.Context, ident iden
 	return nil, berrors.UnauthorizedError("Incorrect TXT record %q%s found at %s",
 		invalidRecord, andMore, challengeSubdomain)
 }
+
+func (va *ValidationAuthorityImpl) validateDNS01(ctx context.Context, ident identifier.ACMEIdentifier, challenge core.Challenge) ([]core.ValidationRecord, error) {
+	if ident.Type != identifier.DNS {
+		va.log.Infof("Identifier type for DNS challenge was not DNS: %s", ident)
+		return nil, berrors.MalformedError("Identifier type for DNS was not itself DNS")
+	}
+
+	// Compute the digest of the key authorization file
+	h := sha256.New()
+	h.Write([]byte(challenge.ProvidedKeyAuthorization))
+	authorizedKeysDigest := base64.RawURLEncoding.EncodeToString(h.Sum(nil))
+
+	// Look for the required record in the DNS
+	challengeSubdomain := fmt.Sprintf("%s.%s", core.DNSPrefix, ident.Value)
+
+	// Return the validation record if the correct TXT record is found
+	return va.validateTXT(ctx, ident, authorizedKeysDigest, challengeSubdomain)
+}
+
+// Compute the DNS-ACCOUNT-01 challenge subdomain per the
+// acme-scoped-dns-challenges specification
+func getDNSAccountChallengeSubdomain(
+	accountResourceURL string,
+	scope core.AuthorizationScope,
+	domain string,
+) string {
+	// "_" || base32(SHA-256(<ACCOUNT_RESOURCE_URL>)[0:10]) || "._acme-" || <SCOPE> || "-challenge"
+	acctHash := sha256.Sum256([]byte(accountResourceURL))
+	acctLabel := strings.ToLower(base32.StdEncoding.EncodeToString(acctHash[0:10]))
+	challengeSubdomain := fmt.Sprintf("_%s._acme-%s-challenge.%s",
+		acctLabel, scope, domain)
+
+	return challengeSubdomain
+}
+
+// validateDNSAccount01 validates a DNS-ACCOUNT-01 challenge using the account's URI
+// (derived from the accountID) and the provided scope.
+func (va *ValidationAuthorityImpl) validateDNSAccount01(ctx context.Context,
+	ident identifier.ACMEIdentifier,
+	challenge core.Challenge,
+	scope core.AuthorizationScope,
+	regid int64,
+) ([]core.ValidationRecord, error) {
+	if ident.Type != identifier.DNS {
+		va.log.Infof("Identifier type for DNS challenge was not DNS: %s", ident)
+		return nil, berrors.MalformedError("Identifier type for DNS was not itself DNS")
+	}
+
+	// Reject unsupported scopes
+	if scope != core.AuthorizationScopeHost && scope != core.AuthorizationScopeWildcard {
+		va.log.Infof("Unsupported scope for DNS-ACCOUNT-01 challenge: %s", scope)
+		return nil, berrors.MalformedError("Unsupported scope for DNS-ACCOUNT-01 challenge")
+	}
+
+	// Compute the digest of the key authorization file
+	h := sha256.New()
+	h.Write([]byte(challenge.ProvidedKeyAuthorization))
+	authorizedKeysDigest := base64.RawURLEncoding.EncodeToString(h.Sum(nil))
+
+	// Iterate over the account URIs to find the correct one
+	// for the accountID and track any errors
+	var validationError error
+	for _, accountURIPrefix := range va.accountURIPrefixes {
```

Comment:
This differs from the latest draft. The recommendation is to use the KID from the JWS message:

https://aaomidi.github.io/draft-ietf-acme-scoped-dns-challenges/#section-4-8.2

> Compute the validation domain name with the KID value in the JWS message

Of course Boulder should be validating that KID value in the JWS message before it's used.

This would remove the need for this for loop too.

Reply:
Although let's probably focus on getting this one done first? https://github.com/letsencrypt/pebble/pull/435

I believe that's the desire from the Let's Encrypt team too?

Reply:
The `kid` from the JWS is parsed in the WFE (as `accountURL`). The `accountID` (aka `regid`) [portion of the URL is retrieved](https://github.com/fastly/boulder/blob/a948e01a300d6b47348c4a750bc410ef73425609/wfe2/verify.go#L465) from the `kid` and stored before [discarding the remainder of the URL](https://github.com/fastly/boulder/blob/a948e01a300d6b47348c4a750bc410ef73425609/wfe2/verify.go#L509-L510).

```golang
        accountURL := header.KeyID
        accountID, prob := wfe.acctIDFromURL(accountURL, request)
```

The `accountID` portion of the accountURL is properly authenticated information, since the entire JWS is validated first and `accountID` is extracted from the protected header.

As the remainder of the `kid` (the `accountURL` prefix) is not passed along from the WFE, we need to reconstruct this in the VA. The existing method used for CAA account binding is to specify the accountURL prefix in a configuration file. The loop is used to allow for migrations (e.g. "old" and "new" prefixes).

IF the `kid` value is required verbatim (versus the reconstituted version currently used), the Boulder changes would be more substantial in order to convey the additional data to the VA.

Reply:
The current method would technically create unwanted behavior, as it would have the server search potentially more than 1 DNS record for the resource, which is an unintended side effect and could potentially be considered non-compliant behavior.

I think this might warrant the extra changes necessary to pass those through here?

Reply:
This implementation could result in two DNS lookups per dns-account-01 challenge during a migration period, e.g. ACMEv1 to ACMEv2. Outside of this period, the Boulder configuration should specify a single value for the ACME account prefix.

Reply:
Boulder currently configures two ACME account URI prefixes: the one from when we were still running acme-v01, and the current acme-v02 prefix. There are still very old, very stable ACME clients which present an "https://acme-v01.api.letsencrypt.org" KID in their JWS headers, and removing that option would break those clients seemingly unnecessarily.

That said, I agree that it's suboptimal for the ACME server to make multiple queries. And if the draft continues forward with the current language, it would be non-compliant to do so: it would be possible for a client to use the acme-v01 KID in their request JWS, and then for us to confirm the presence of a DNS record computed from an acme-v02-prefixed KID.

So one path forward that we're considering (and that @jsha and I discussed earlier today) is this: only allow a single accountURIPrefix (which would be the acme-v02 one) to be configured for the purposes of this challenge. When we receive a "please validate this challenge for me" request, confirm that the KID in the request also has that same singular prefix; if not, reject the request.

But teaching the WFE about one blessed account prefix and giving it extra challenge-specific logic doesn't sound that much simpler than simply plumbing the request's KID through into the PerformValidation request. So I'm personally also considering that path forward.

One other path forward would be to update the draft: one possibility would be to update the draft to say that the validation domain should be computed from the value returned by the `Location:` header in the Server's response to a new-acct `onlyReturnExisting` request.

Reply:
It seems straightforward enough to append the KID to the ra.PerformValidationRequest and va.PerformValidationRequest protobufs. Any reasons against plumbing it through?

Reply:
Another approach would be to attach the AccountKeyID to the Authz when the WFE sends to the RA. This would flow through to the RA->VA as well and be available for validation.

Reply:
This last approach doesn't match the Challenge object description
```
// Challenge is an aggregate of all data needed for any challenges.
//
// Rather than define individual types for different types of
// challenge, we just throw all the elements into one bucket,
// together with the common metadata elements.
```
by putting the KID in the Authorization object. However it seems to fit more naturally as being scoped to the entire Authorization.

Reply:
As there is discussion on IETF ACME working group about how to tell witch parental scope client want to use: like to challagne on example.com for get a cert for foo.example.com, I think we'll need plumbing from WFE to VA anyway, if we make pipe one can add everything needed from it from accounturi client sent to what scope clinet want to use
https://mailarchive.ietf.org/arch/msg/acme/HiD1xNZD7b4kLgHmYAsmVuT0HVA/

Reply:
Right. I was thinking about moving the `Scope` attribute from the Challenge to the Authorization as well for the same reason.

Reply:
In the [latest change](https://github.com/letsencrypt/boulder/pull/7387/commits/a6f87c4e4eb8def1058e06d35db9c46d9d3adf4c), the Authorization.AccountURL is populated in the WFE with the JWS `kid` and passed via the RA to the VA. The Authorization.Scope is currently unused (and removed before sending to clients) but can be used for passing scope to/from client.

The scope decision currently rests with the VA, but this could move to RA and be potentially modified by the client requested scope.